File: langs/dotnet/src/Dotlyte/Merger.cs
namespace Dotlyte;

/// <summary>
/// Deep merge utility.
/// </summary>
public static class Merger
{
    /// <summary>
    /// Deep merge two dictionaries. Override values win.
    /// </summary>
    public static Dictionary<string, object?> DeepMerge(
        Dictionary<string, object?> baseDict,
        Dictionary<string, object?> overrideDict)
    {
        var result = new Dictionary<string, object?>(baseDict);

        foreach (var (key, value) in overrideDict)
        {
            if (result.TryGetValue(key, out var existing)
                && existing is Dictionary<string, object?> existingDict
                && value is Dictionary<string, object?> overrideInner)
            {
                result[key] = DeepMerge(existingDict, overrideInner);
            }
            else
            {
                result[key] = value;
            }
        }

        return result;
    }
}

File: langs/php/src/Merger.php
<?php

declare(strict_types=1);

namespace Dotlyte;

/**
 * Deep merge utility.
 */
final class Merger
{
    /**
     * Deep merge two associative arrays. Override values win.
     *
     * @param array<string, mixed> $base
     * @param array<string, mixed> $override
     * @return array<string, mixed>
     */
    public static function deepMerge(array $base, array $override): array
    {
        $result = $base;

        foreach ($override as $key => $value) {
            if (
                isset($result[$key])
                && is_array($result[$key])
                && is_array($value)
            ) {
                $result[$key] = self::deepMerge($result[$key], $value);
            } else {
                $result[$key] = $value;
            }
        }

        return $result;
    }
}
